商品出品機能

k-foo opened the pull request:
# What
- 商品出品機能の実装（ActiveHash・Active Storage・JavaScriptの活用）
- モデル単体テストの実装
# Why
- ユーザーが、販売したい商品の登録をするため
- データベースに保存する際、保存できないデータを入力していないか確認をするため。

【確認動画について】
●ユーザー登録している人が出品できる
https://gyazo.com/225b707493b21f41b3d383e05ee1be21

●ユーザー登録していない人は出品できない（ログイン画面へ）
https://gyazo.com/177ece22fe79020cf5eeca371b31d718

●必要事項を入力していないと出品できない（エラーが出力される）※価格は空欄だった時でテスト
https://gyazo.com/4565c2f107a2dd88f2e363bc92a22dab

●ActiveStorageの実装確認（画像が添付できるか）
https://gyazo.com/76c46bb0a5092bb1050dedb8e50cef70

●JavaScriptの実装確認（手数料・利益計算できるか）
https://gyazo.com/94fe0667c15defcea3455a518dba7aff

●ActiveHashの実装確認（きちんと表示されるか）
前編
https://gyazo.com/a4ed3eed5df050a6162a24c417511a05
後編
https://gyazo.com/3b2cdbec4258de099f3b47795a7c752e

●必要事項を入力していると出品できる（データベースへ保存される）
https://gyazo.com/a5dcff7c55dd885602c5446bee532428

●モデル単体テストコードの実行結果（`bundle exec rspec spec/models/item_spec.rb`）
https://gyazo.com/73de741bc4a65496002b6c9589208383

k-foo committed:
itemモデルの作成

rails g model itemにて作成。

k-foo committed:
商品出品情報を保存するためのテーブル作成。

rubocopを実行したため、user機能の体裁微修正も発生した。

k-foo committed:
ActiveHashの実装。

データベースを作成しないカテゴリー、商品の状態、配送料の負担、発送元の地域、発送までの日数についてActiveHashを活用した。

k-foo committed:
mini_magick、image_processing、active_hash（Gem）の追加。

k-foo committed:
itemのビューファイルについてname属性等を修正。

カラムに合わせたname属性・ActiveHashにて作成したクラス名・エラー発生時にメッセージが表示されるよう実装。

k-foo committed:
itemモデルとuserモデルについて修正。

アソシエーションを実装したり、itemモデルに関してはバリデーションとActiveHashのアソシエーションも実装した。

k-foo committed:
active_storageのマイグレーションファイルをDBへ反映。

Gem導入とともに、自動生成されたマイグレーションファイルをDBへ反映。

k-foo committed:
出品ボタンを押した時に出品登録画面へ遷移するよう実装。

k-foo committed:
shipping_costs_idカラムの修正。

モデルは単数形で、itemのデータベース（マイグレーションファイルは複数形）で作成してしまっていたためデータが保存できず、ロールバックにて修正を行った。（単数形に統一）それに伴って、READMやER図も修正。

k-foo committed:
イメージ画像（テスト用）を実装。

k-foo committed:
itemモデルテストコードの実装。

ダミーデータとしてFakerを使用。

k-foo committed:
エラー解消のため一行追加。（AcriveStrage×Rspec）

Rspecを使って、テストコードを実装する際、ActiveStrageによるエラーが発生した。Rspesのイシューで解決方法を見つけたため、実装。（→エラー解消）

k-foo committed:
バリデーションに「:image」追加。

テストコード実装時に、エラーが拾えず、バリデーションが足りなかったことに気がついたため。

k-foo committed:
販売手数料・利益をJavaScriptで実装。

chirikuzu commented on app/controllers/items_controller.rb:
```diff
@@ -3,5 +3,25 @@ class ItemsController < ApplicationController
   before_action :authenticate_user!, except: :index
 
   def index
+    @items = Item.order('created_at DESC')
```
```ruby
@items = Item.order('created_at DESC')
```
こちらのプルリクエストは、「商品出品機能の実装」でお間違いないでしょうか。
商品一覧表示機能に関係するメソッドやアクションについて、このプルリクエストでレビューしてしまうと、今後のプルリクエストで拝見したいファイルが差分として上がってこない可能性があります。
商品一覧表示機能に関係するメソッドは、次回以降のプルリクエストで拝見させていただきたいので、一旦削除もしくはコメントアウトをお願いいたします。

k-foo replied:
修正しました！（コメントアウトしました）

chirikuzu commented on app/models/item.rb:
```diff
@@ -0,0 +1,42 @@
+    validates :delivery_area_id
+    validates :shipping_date_id
+    # 300円以上かつ半角数字でないと入力不可
+    validates :price, numericality: { greater_than_or_equal_to: 300 }
```
```ruby
validates :price, numericality: { greater_than_or_equal_to: 300 }
```
こちら9999999以下でないと出品ができないようにバリデーションを追加しましょう。

k-foo replied:
追加実装しました！

chirikuzu commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,91 @@
+        @item.price = 100
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Price must be greater than or equal to 300')
+      end
```
10000000円以上で出品ができないことについてのテストを追加しましょう。

k-foo replied:
テストしました！

chirikuzu reviewed (commented):
レビュー依頼ありがとうございます。
コメントしましたので、ご確認をお願いいたします。

erika618 commented:
10000000円以上で出品ができないことについての確認動画
【10000000円以上で出品ができないことについてのテストコード実行】
https://gyazo.com/03582f7c8faf278e8018c83eb9e45456

【商品出品画面できちんとエラーが出ることの確認】
https://gyazo.com/a28b313303567ab03c710081f795f93d

k-foo committed:
10000000円以上で出品ができないように実装。

NaturalRadio commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,96 @@
+        expect(@item.errors.full_messages).to include("Description can't be blank")
+      end
+      it 'カテゴリーの情報が「---」だと出品できない' do
+        @item.category_id = 0
```
〇〇_idに対して`presence: true`の記述がありますので
```
〇〇_idが空の場合だと出品できない
```
こちらのテスト項目を追加してみましょう！

k-foo replied:
修正しました！
【実装後のテストコードの動画】
https://gyazo.com/c96f3065699428d0ec2d1609e8621449

プッシュし忘れていました。申し訳ありません。

NaturalRadio reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします！

k-foo committed:
itemモデルの単体テストコードの修正。（user_idを必須とする）

chirikuzu commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,101 @@
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Description can't be blank")
+      end
+      it 'カテゴリーの情報が「---」だと出品できない' do
+        @item.category_id = 0
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Category must be other than 0')
+      end
+      it '商品の状態の情報が「---」だと出品できない' do
+        @item.item_status_id = 0
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Item status must be other than 0')
+      end
+      it '配送料の負担の情報が「---」だと出品できない' do
+        @item.shipping_cost_id = 0
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Shipping cost must be other than 0')
+      end
+      it '発送元の地域の情報が「---」だと出品できない' do
+        @item.delivery_area_id = 0
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Delivery area must be other than 0')
+      end
+      it '発送までの日数の情報が「---」だと出品できない' do
+        @item.shipping_date_id = 0
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Shipping date must be other than 0')
+      end
```
> 〇〇_idに対してpresence: trueの記述がありますので
> 
> ```
> 〇〇_idが空の場合だと出品できない
> ```
> 
> こちらのテスト項目を追加してみましょう！

こちら意図を正確にお伝えできず大変失礼いたしました。
商品の状態, 配送料の負担, 発送元の地域,  発送までの日数, カテゴリーのカラムに対してpresence: trueを設定していただいておりますので、こちらそれぞれが空の場合に出品ができないことについてもテストを追加しましょう。

k-foo replied:
恐れ入ります。
それらは、最初のプッシュの際に行ったという認識だったのですが
（それも含めてのテストコードの実装だと思っているのですが）どういう意味でしょうか。

【コード前半】
https://gyazo.com/4328d612ae579660440ca9461c30335e
【コード中盤】
https://gyazo.com/6bc6fdc99ccdefdf58548679c37c51a9
【コード後半】
https://gyazo.com/3bb2842ca82a1503a30cb1b8a493f401

「---」の時には出品できない（何も選択していない）という認識なのですが、誤っていますでしょうか。
それとも、モデルの単体テストコードではなく、一連の流れの方のテストを指していらっしゃいますか？

wakita0217 replied:
ご返信ありがとうございます。
> 「---」の時には出品できない（何も選択していない）という認識なのですが、誤っていますでしょうか。

仰る通り現在モデルファイルに`numericality: { other_than: 0 }`がありますので、例えばcategoryが0（---）の時には出品できない形になります。
加えて現在presence :trueもあるため、categoryはnilの時にも出品できません。
上記二つの挙動を確認するために以下テストを**両方**作成していただければと思います。
・‘カテゴリーが0（---）だと出品できない’
・‘カテゴリーが空だと出品できない’

（出品時にカテゴリーをnilにする選択肢はありませんが、chromeの検証ツールなどを使用すればユーザ側でカテゴリーの選択欄を無くしてカテゴリーの値をnilで送信することができてしまいます。そういった操作に備えるためにpresence :trueと、対応したテストを作成していただければと思います）

またカテゴリ以外（商品の状態, 配送料の負担, 発送元の地域, 発送までの日数）についても同様です。

k-foo replied:
ご丁寧にありがとうございます。空欄にして入力できてしまうのですね・・！
知らなかったのでとても勉強になりました！！
修正したので、よろしくお願いします。

【実装後のテストコードの実行動画】
https://gyazo.com/3e48bd50d9fc051e5500a655b20c37b9


※合わせて、「nil」と「’’」の違いってなんだろうと調べました。
https://qiita.com/somewhatgood@github/items/b74107480ee3821784e6
●nil（多言語だとNULL）‥「何も存在していない」
●ブランク（””や’’）..「値を入れる箱は存在しているが、中身は無い」
●スペース..「値が定義されており、半角、全角も区別する」
  今回のでいうとnilが適切？と思い、合わせて修正してみました。よろしくお願いします。

chirikuzu reviewed (commented):
レビュー依頼ありがとうございます。
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします。

wakita0217 reviewed (commented):
レビュー依頼ありがとうございます。
頂いたコメントに返信いたしましたので、ご確認をお願いいたします。

k-foo committed:
itemモデルの単体テストコードの修正。（空欄もNGとする）

ActiveHashについて、選択肢が「０」だと保存できないのと同じように「nil」であっても保存できないことをテストコードとして実装。合わせて’’をnilに修正。

s-aka-moto reviewed (commented):
レビュー依頼ありがとうございます！
LGTMとさせていただきます！実装お疲れ様でした！
このブランチをマージし、masterブランチでプルを行いましょう！
機能が反映されていることを確認し、引き続き実装頑張っていきましょう！
